app/vmalert/rule/alerting.go

```diff
@@ -229,6 +208,50 @@ func (ar *AlertingRule) updateWith(r Rule) error {
 	return nil
 }
 
+// loadAlert returns copied alert if id exists
+func (ar *AlertingRule) loadAlert(id uint64) *notifier.Alert {
+	activeAlerts := ar.alerts.Load()
+	if activeAlerts == nil {
+		return nil
+	}
+	for k, v := range *activeAlerts {
+		if k == id {
+			na := *v
+			return &na
+		}
+	}
+	return nil
+}
+
+// loadAlerts returns copied alerts
+func (ar *AlertingRule) loadAlerts() map[uint64]*notifier.Alert {
+	activeAlerts := ar.alerts.Load()
+	if activeAlerts == nil {
+		return map[uint64]*notifier.Alert{}
+	}
+	newAlerts := make(map[uint64]*notifier.Alert)
+	for k, v := range *activeAlerts {
+		na := *v
```

Comment:
This can copy only primitive types... more complex types like `map` will keep the old reference and may backfire in the future.

Reply:
I think, Atomic.Pointer is right direction for potential fix. 

 I took a brief look at vmalert code. And it seems, it may have a potential data races, since not all calls for ar.alerts are protected with mutex lock. Such as, alertsToTimeseries or execRange.

 Maybe, it'll be enough to introduce Clone() method for notifier.Alert. It must properly clone all non primitive types. 

It should reduce number of locks and simplify code a bit.

Reply:
>Such as, alertsToTimeseries or execRange.

After check, there should be no data race on alerts now. `alertsToTimeseries` is under protected and `execRange` won't be execute concurrently by design. 
We still stick to mutex lock now, might change to atomic value in future refactor as it can simplify the code.